### ISSUE TITLE
Nix store doesn't get mounted on MacOS

### ISSUE THREAD

**Issue by jheidbrink** [NixOS/nix#6012]
**Describe the bug**

When installing Nix on MacOS 12.1 on a Macbook Air with M1 chip (for the second time as after the first install and reboot Nix commands were not available), the installer got stuck at

```
---- sudo execution ------------------------------------------------------------
I am executing:

    $ sudo launchctl kickstart -k system/org.nixos.darwin-store

to launch the Nix volume mounter
```

(like in issue #5547)

I found that the installer was repeatedly calling diskutil, and opening Disk Utility I found that the Nix store was unmounted. After mounting it manually, the installer proceeded and finished successfully. After rebooting, the Nix store is unmounted and the nix commands are not available. Trying to mount the disk manually with Disk utility shows an error dialog

> Could not mount “Nix Store”. (com.apple.DiskManagement.disenter error 49218.)

**Additional context**

@abathur mentioned that my mounting issue is reported a lot on Mac Forums. Several users reported that killing fsck resolved the issue for them, but a `pstree | grep -B 5 fsck` revealed no fsck running on my machine. I won't have time to further investigate today, but update this once I find anything useful.

@abathur suggested the output of the following commands could be helpful:

<details>
  <summary>Click to expand</summary>

```
/usr/sbin/diskutil mount -mountPoint "/nix" EA0C28DB-068C-4BC4-8CCC-EE795A49A1F7; echo $?
Mountpoint /nix does not exist
1

sudo mkdir /nix
mkdir: /nix: Read-only file system
```

```
/usr/sbin/diskutil info disk3s7; echo $?                                                 
   Device Identifier:         disk3s7
   Device Node:               /dev/disk3s7
   Whole:                     No
   Part of Whole:             disk3

   Volume Name:               Nix Store
   Mounted:                   No

   Partition Type:            41504653-0000-11AA-AA11-00306543ECAC
   File System Personality:   APFS
   Type (Bundle):             apfs
   Name (User Visible):       APFS
   Owners:                    Disabled

   OS Can Be Installed:       No
   Booter Disk:               disk3s2
   Recovery Disk:             disk3s3
   Media Type:                Generic
   Protocol:                  Apple Fabric
   SMART Status:              Verified
   Volume UUID:               EA0C28DB-068C-4BC4-8CCC-EE795A49A1F7
   Disk / Partition UUID:     EA0C28DB-068C-4BC4-8CCC-EE795A49A1F7

   Disk Size:                 245.1 GB (245107195904 Bytes) (exactly 478724992 512-Byte-Units)
   Device Block Size:         4096 Bytes

   Container Total Space:     245.1 GB (245107195904 Bytes) (exactly 478724992 512-Byte-Units)
   Container Free Space:      177.2 GB (177211043840 Bytes) (exactly 346115320 512-Byte-Units)

   Media OS Use Only:         No
   Media Read-Only:           No
   Volume Read-Only:          Not applicable (not mounted)

   Device Location:           Internal
   Removable Media:           Fixed

   Solid State:               Yes
   Hardware AES Support:      Yes

   This disk is an APFS Volume.  APFS Information:
   APFS Container:            disk3
   APFS Physical Store:       disk0s2
   Fusion Drive:              No
   FileVault:                 Yes
   Sealed:                    No
   Locked:                    No

0
```

```
/usr/sbin/diskutil apfs list
APFS Containers (3 found)
|
+-- Container disk3 E7ABBD36-2413-47D1-B168-AC3A2406B91E
    ====================================================
    APFS Container Reference:     disk3
    Size (Capacity Ceiling):      245107195904 B (245.1 GB)
    Capacity In Use By Volumes:   67898703872 B (67.9 GB) (27.7% used)
    Capacity Not Allocated:       177208492032 B (177.2 GB) (72.3% free)
    |
    +-< Physical Store disk0s2 89FB8B6B-215C-4CED-BCE9-045DBCF4867D
    |   -----------------------------------------------------------
    |   APFS Physical Store Disk:   disk0s2
    |   Size:                       245107195904 B (245.1 GB)
    |
    +-> Volume disk3s1 3EB6D93A-00BA-4461-A1FC-23788A35E0F5
    |   ---------------------------------------------------
    |   APFS Volume Disk (Role):   disk3s1 (System)
    |   Name:                      Macintosh HD (Case-insensitive)
    |   Mount Point:               Not Mounted
    |   Capacity Consumed:         15752564736 B (15.8 GB)
    |   Sealed:                    Broken
    |   FileVault:                 Yes (Unlocked)
    |   Encrypted:                 No
    |   |
    |   Snapshot:                  A4DC6984-7204-4D8C-90D0-62A925AA35A4
    |   Snapshot Disk:             disk3s1s1
    |   Snapshot Mount Point:      /
    |   Snapshot Sealed:           Yes
    |
    +-> Volume disk3s2 48F77FF2-C341-4408-83FD-BA57901E712A
    |   ---------------------------------------------------
    |   APFS Volume Disk (Role):   disk3s2 (Preboot)
    |   Name:                      Preboot (Case-insensitive)
    |   Mount Point:               /System/Volumes/Preboot
    |   Capacity Consumed:         413794304 B (413.8 MB)
    |   Sealed:                    No
    |   FileVault:                 No
    |
    +-> Volume disk3s3 02B6E604-703E-4086-8C17-A60215E645AA
    |   ---------------------------------------------------
    |   APFS Volume Disk (Role):   disk3s3 (Recovery)
    |   Name:                      Recovery (Case-insensitive)
    |   Mount Point:               Not Mounted
    |   Capacity Consumed:         844795904 B (844.8 MB)
    |   Sealed:                    No
    |   FileVault:                 No
    |
    +-> Volume disk3s5 6F272577-3946-4D2C-BA3F-8161DC5A1F1A
    |   ---------------------------------------------------
    |   APFS Volume Disk (Role):   disk3s5 (Data)
    |   Name:                      Data (Case-insensitive)
    |   Mount Point:               /System/Volumes/Data
    |   Capacity Consumed:         47128596480 B (47.1 GB)
    |   Sealed:                    No
    |   FileVault:                 Yes (Unlocked)
    |
    +-> Volume disk3s6 0670E08D-A6DF-4732-AA7E-9E84E11BD082
    |   ---------------------------------------------------
    |   APFS Volume Disk (Role):   disk3s6 (VM)
    |   Name:                      VM (Case-insensitive)
    |   Mount Point:               /System/Volumes/VM
    |   Capacity Consumed:         20480 B (20.5 KB)
    |   Sealed:                    No
    |   FileVault:                 No
    |
    +-> Volume disk3s7 EA0C28DB-068C-4BC4-8CCC-EE795A49A1F7
        ---------------------------------------------------
        APFS Volume Disk (Role):   disk3s7 (No specific role)
        Name:                      Nix Store (Case-insensitive)
        Mount Point:               Not Mounted
        Capacity Consumed:         3618881536 B (3.6 GB)
        Sealed:                    No
        FileVault:                 Yes (Unlocked)
```
</details>

**Comment by abathur** [comment]
This is actually a good sign of sorts. I am not sure why, but the fact that your /nix mount point doesn't exist suggests that your /etc/synthetic.conf is either missing or doesn't list nix.

I would check on the status of that file, and maybe see if there were any problems reported with that part of the install process.

**Comment by jheidbrink** [comment]
> the fact that your /nix mount point doesn't exist suggests that your /etc/synthetic.conf is either missing or doesn't list nix.

That's correct, and in my case it also mentions that the file is autogenerated and any changes to it will be overwritten, thus explaining why the `/nix` mountpoint was missing.. I have not yet clarified if I'm allowed to change that (I don't own the laptop). Anyway, it might be a bit out of scope for the Nix installer to deal with situations like these, so this issue can probably be closed.

One thing that puzzles me is why the installer got stuck and why manually mounting the volume helped. I guess the `/nix` folder must have been present when I clicked the mount button in order for the mounting to succeed, so why couldn't the installer mount it itself?

**Comment by abathur** [comment]
Hmm. Maybe out of scope to deal with it cleanly, but it would be nice if there's a way to fail more transparently, regardless of when... 

> One thing that puzzles me is why the installer got stuck 

The small/tautological reason is that it got stuck because the volume didn't mount. This part of the installer could be a little smarter (maybe with a timeout, and some debug print/exit when reached), but it's stuck in a loop awaiting the mount. 

I'm much less sure about the ~real reason it got stuck: whatever reason it failed to mount. The main reason we've caught so far is that the volume was already mounted elsewhere, but the installer now attempts to prophylactically force an unmount before this section to avoid that case. If the problem reliably happens on a from-scratch install, it might help to monitor `log` or `Console.app` during the install and sift for anything that clarifies. (It would be great to know, if you can reliably reproduce the case and it isn't too much trouble--but no worries if you can't.)

> and why manually mounting the volume helped. I guess the /nix folder must have been present when I clicked the mount button in order for the mounting to succeed, so why couldn't the installer mount it itself?

Speculating a bit, but my guess would be that you aren't prevented from modifying the file, but it does get re-set by some process. When the installer runs, it updates the file and then immediately calls a system utility (apfs.util) that creates the mount point. But if something else removes/overwrites the file, the mountpoint won't be present after a reboot.

One other thing worth a try: starting with Big Sur+, macOS also supports an `/etc/synthetic.d/<file>` form. You could create a file `/etc/synthetic.conf/nix` containing "nix" on a line by itself, reboot, and see if the system creates the mountpoint. (Whatever overwrites synthetic.conf may already account for this--or might suddenly start doing so in the future, of course.)

**Comment by jheidbrink** [comment]
> If the problem reliably happens on a from-scratch install

It actually doesn't, the first Nix installation finished without problems. I still have a copy of the console output of that installation, and going through it I can see nothing unusual. I only encountered this problem when I rebooted, discovered that Nix commands were not available and attempted another installation.

btw in the meantime I clarified that I am allowed to modify the autogenerated `/etc/synthetic.conf`, and appending a line "nix" there helped.
